core/src/main/java/org/jdbi/v3/CollectorFactoryRegistry.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3;
+
+import org.jdbi.v3.tweak.CollectorFactory;
+
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.function.Function;
+import java.util.stream.Collector;
+
+/**
+ * Registry of collector factories.
+ * Contains a set of collector factories, registered by the application.
+ */
+class CollectorFactoryRegistry {
+
+    private final Set<CollectorFactory<?, ?, ?>> factories = new CopyOnWriteArraySet<>();
+
+    CollectorFactoryRegistry() {
+        factories.add(new SortedSetCollectorFactory());
+    }
+
+    CollectorFactoryRegistry(CollectorFactoryRegistry parent) {
+        factories.addAll(parent.factories);
+    }
+
+    CollectorFactoryRegistry createChild() {
+        return new CollectorFactoryRegistry(this);
+    }
+
+    void register(CollectorFactory<?, ?, ?> factory) {
+        factories.add(factory);
+    }
+
+    @SuppressWarnings("unchecked")
+    <T, R> Collector<T, ?, R> createCollectorFor(Class<R> type) {
```

Comment:
I know you are only mirroring how the existing registries work, but I wonder if it would make sense for these to either return a nullable type or an `Optional<Collector<...>>` so that lookups can be done without necessarily throwing an exception.  If so, probably a change that should be done to the existing registries as well.